### PR TITLE
feat(IoTDA) add a new resource to manage device

### PR DESCRIPTION
**Pull request by hmbig2**
**What this PR does / why we need it**:
add a new resource to manage device

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

1. An API error occurred while setting the shadow, so don't support shadow in this PR.

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/iotda' TESTARGS='-run=TestAccDevice_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/iotda -v -run=TestAccDevice_basic -timeout 360m -parallel 4
=== RUN   TestAccDevice_basic
--- PASS: TestAccDevice_basic (29.52s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/iotda     29.632s
```

### CODE REVIEW

File: docs/resources/iotda_device.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the IoTDA device resource.
+If omitted, the provider-level region will be used. Changing this parameter will create a new resource.
+
+* `name` - (Required, String) Specifies the device name, which contains 4 to 256 characters. Only letters,
+Chinese characters, digits, hyphens (-), underscore (_) and the following specail characters are allowed: `?'#().,&%@!`.
+
+* `device_id` - (Optional, String, ForceNew) Specifies the device ID, which contains 4 to 256 characters.
+Only letters, digits, hyphens (-) and underscore (_) are allowed. If omitted, the platform will automatically allocate
+a device ID. Changing this parameter will create a new resource.
```
Comment (Jason-Zhang9309) [inline docs/resources/iotda_device.md]: move this optional parameter to line 62

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+				Computed:      true,
+				ConflictsWith: []string{"secret"},
+				ValidateFunc: func(i interface{}, k string) (warnings []string, errors []error) {
+					v, ok := i.(string)
+					if !ok {
+						errors = append(errors, fmt.Errorf("expected type of %s to be string", k))
+						return warnings, errors
+					}
+
+					if len(v) == 40 || len(v) == 64 {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: should be `if len(v) != 40 && len(v) != 64`?

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+	}
+
+	if resp.DeviceId == nil {
+		return diag.Errorf("error creating IoTDA device: id is not found in API response")
+	}
+
+	d.SetId(*resp.DeviceId)
+
+	// bind tags
+	err = bindDeviceTags(client, d.Id(), nil, d.Get("tags").(map[string]interface{}))
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go:182]: add a check of whether to tags is empty

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+	if err != nil {
+		return diag.Errorf("error binding tags when creating IoTDA device: %s", err)
+	}
+
+	// freeze device
+	isFrozen := d.Get("frozen").(bool)
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go:188]: add a check too

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	// name,desc
+	updateOpts := buildDeviceUpdateParams(d)
+	_, err = client.UpdateDevice(updateOpts)
+
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: extra blank line

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+	isFrozen := d.Get("frozen").(bool)
+	err = updateDeviceStatus(client, d.Id(), utils.StringValue(resp.Status), isFrozen)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return ResourceDeviceRead(ctx, d, meta)
+}
+
+func ResourceDeviceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: s/ctx/_

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+		err = bindDeviceTags(client, d.Id(), o.(map[string]interface{}), n.(map[string]interface{}))
+		if err != nil {
+			return diag.Errorf("error updating the tags of IoTDA device: %s", err)
+		}
+	}
+
+	return ResourceDeviceRead(ctx, d, meta)
+}
+
+func ResourceDeviceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: s/ctx/_

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+		d.Set("description", response.Description),
+		d.Set("space", response.AppId),
+		d.Set("status", response.Status),
+		d.Set("node_type", response.NodeType),
+		d.Set("tags", flattenTags(response.Tags)),
+		d.Set("frozen", utils.StringValue(response.Status) == deviceStatusFrozen),
+	)
+
+	if response.AuthInfo != nil {
+		mErr = multierror.Append(
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: please use `mErr = multierror.Append(mErr,`

Comment (hmbig2) [reply huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: ok

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go (hunk 2 above)
Comment (hmbig2) [reply huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: fix, and add the test

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	v5 "github.com/huaweicloud/huaweicloud-sdk-go-v3/services/iotda/v5"
```
Comment (ShiChangkuo) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go:12]: the package name already is **v5**, alias is unnecessary.

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+func ResourceDeviceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	// name,desc
+	updateOpts := buildDeviceUpdateParams(d)
```
Comment (ShiChangkuo) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: should add d.HasChanges("name", "description") check

Comment (hmbig2) [reply huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: done

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go (hunk 9 above)
Comment (hmbig2) [reply huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go:12]: It's auto-format by vscode.

**Pull request merged**